Multiple `ScrollingHitObjectContainer` logic simplification

ekrctb opened the pull request:
There are more possibilities of code shrinking (simplifying `switch` statements) but I didn't include those because I want to make a separate PR that contains only syntactical (no semantical) changes.

1. Don't recurse into nested hit objects in `OnAdd` and `OnRemove`.

Only top-level hit objects were/are checked for layout computation caching.
Also, the lifetime of nested hit objects is not managed by the HitObjectContainer.

2. Remove delay of lifetime update.

Because of changes of `HitObjectContainer` (`OnAdd` timing for non-pooled DHOs), it is no longer necessary to manage a set of hit objects that needs lifetime update.

3. Merge duplicated scrolling length computation into one, removing ad-hoc caching of this value.

No noticeable performance difference.

ekrctb committed:
Don't add nested hit objects to the sets

Only top-level hit objects are checked for layout computation caching.
Also, lifetime of nested hit objects are not managed by the HitObjectContainer.

ekrctb committed:
Remove delay of lifetime update

The scheduling is no longer necessary because `OnAdd` is changed to not invoked immediately for non-pooled DHOs.

ekrctb committed:
Remove `scrollLength` caching field

It was not clear when the field is updated.

smoogipoo commented:
This looks like it's regressed the origin adjusted start time somehow.

Master:
https://drive.google.com/file/d/1S-Q15qAstI9j4vDTlDAUqod4fB7ODNqI/view?usp=sharing

This branch:
https://drive.google.com/file/d/1_O41whLdzUChkCUa04M9W6OQaLHWc_gS/view?usp=sharing

Note how the objects pop in on this branch.

ekrctb commented:
The origin adjusted lifetime regression is because taiko hit object uses relative draw size to compute `OriginPosition`. It is invalid until the DHO is added to the container.
A possible solution is to change timing of `OnAdd` so `OnAdd` is invoked for DHOs after added to the container. I think this behavior makes sense in isolation (more useful). There are no other usage of the method in lazer or public custom rulesets.
Do you think this solution is okay?

ekrctb committed:
Change timing of `HitObjectContainer.OnAdd`/`OnRemove`

Now, `OnAdd` is always invoked after the DHO is added to the container.
This change is required by `ScrollingHitObjectContainer` to compute origin adjusted lifetime when origin position is relative to the parent drawable size.

smoogipoo commented:
I think that change looks fine in isolation.

We should also be able to also give taiko a valid size by calculating it as `TaikoHitObject.DEFAULT_SIZE * TaikoPlayfield.DEFAULT_HEIGHT` at some point. This is actually done in a lot of places anyway >_>.
